### ISSUE TITLE
Support Environment Addons

### ISSUE THREAD

**Issue by Lou1415926** [aws/copilot-cli#4219]
### Problem Statement
Today, Copilot supports workload-level addons. With this feature, user can add to a service/job [additional AWS resources](https://aws.github.io/copilot-cli/docs/developing/addons/modeling/) that are not integrated to a service/job manifest by default. 

This works nice for resources that are meant to be tied to a single service.  However, there are resources such as AppMesh or Load Balancers that are designed to be shared by multiple consumers; and resources whose lifecycle should not be tied to that of a single service, but rather an environment. 

To address this, we propose the environment addons feature - a user experience for associating additional AWS resources with environments and hence across workloads.

### Proposal

The proposal can be broken down into three parts: 
1. Deploying environment addons 
2. Using values from environment addons in a workload-level addon 
3. Using values from environment addons in a service/job.

#### Deploying Environment Addons

The experience to deploy environment addons will be mirroring that of the existing workload-level addons:

1. Users will write addon templates as a CloudFormation template in YAML format.
2. Users can [customize the `Parameters` section](https://aws.github.io/copilot-cli/docs/developing/addons/modeling/#customizing-the-parameters-section).
5. Users will then store the addon templates in `copilot/environments/addon/myaddon.yml` folder.
6. User will run `copilot env deploy` to deploy an environment along with the addons.
7. `copilot env package --upload-assets` will upload the addon templates. User will be able to deploy environment addons in their pipelines.

##### File Structure

To elaborate on 2. here is an example. Suppose we have a "test" environment, a "dev" environment and a "web" service. Now we want to create an `AppMesh` addon and a few `WAF` rules on environment level:
```shell
copilot/
├── environments/ 
│   ├── addons/                # <- You can define multiple addons. 
│   │     ├── appmesh.yml      # <- The `AppMesh` addon.
│   │     └── waf.yml          # <- The WAF rules.  
│   ├── test/ 
│   │     └─── manifest.yml 
│   └── dev/ 
│         └── manifest.yml
└── web
    ├── addons/
    │     └── appmesh_node.yml
    └─── manifest.yml 
  ```

All environments (in this example, both "test" and "dev") will share the same addon templates. Just like today’s workload-level addons, any environment-specific configuration should be specified inside of the addon templates via the `Conditions` and `Mappings` sections. This follows [CFN’s best practice of reusing templates](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/best-practices.html#reuse):
```yaml
AWSTemplateFormatVersion: 2010-09-09
Parameters:
  App:
    Type: String
    Description: Your application name.
  Env:
    Type: String
    Description: The environment name your service, job, or workflow is being deployed to.
Conditions:
  IsTestEnv: !Equals [ !Ref Env, "test" ]  # <- Use `Conditions` section to specify "test"-specific configurations.
```

#### Using Values From Environment Addons In A Workload-Level Addon

To use a value from an environment addon in a service/job addon, user needs to export that value from the environment addon stack:
```yaml
# In appmesh.yml
AWSTemplateFormatVersion: 2010-09-09
Parameters:
  App:
    Type: String
    Description: Your application name.
  Env:
    Type: String
    Description: The environment name your service, job, or workflow is being deployed to.
Resources:
    MyAppMesh:
        Type: AWS::AppMesh::Mesh
        Properties: 
            MeshName: !Sub ${Env}-MyAppMesh
Outputs:
  MyAppMeshName:
     Value: !GetAtt AppMesh.MeshName
     Export: 
        Name: !Sub ${App}-${Env}-MyAppMeshName # <- Exported as `app-test-MyAppMeshName` in "test" environment, `app-prod-MyAppMeshName` in "prod" environment.
```
To use the value of `MyAppMeshName` in a service addon, such as `copilot/web/addons/appmesh_node.yml`, users can write an import statement:
```yaml
# In a service addon CloudFormation Stack.
Parameters:
    Env: !Ref Environment
    
Resources:
    APIVirtualNode:
        Type: AWS::AppMesh::VirtualNode
        Properties: 
        MeshName: 
            Fn::ImportValue: 
                !Sub "${App}-${Env}-MyAppMeshName" # <- Imported from the `appmesh.yml` environment addon.
```

#### Using Values From Environment Addons In A Workload (Service/Job)

This works in conjunction with the proposal of [Import Values From Other CFN Stacks in Workload Manifests #4220 ](https://github.com/aws/copilot-cli/issues/4220) to provide a full story of creating additional environment-level resources. Users can use `from_cfn` to reference values exported from the environment addons. Read [Import Values From Other CFN Stacks in Workload Manifests #4220](https://github.com/aws/copilot-cli/issues/4220) for more details!

### Related Issues
#1975, #3112, [#3936](https://github.com/aws/copilot-cli/issues/3936#issuecomment-1249655178)

**Comment by matthewhembree** [comment]
I guess I'm interested in sequencing of addons deployment.

For instance, when I deploy a service with addons, the addons are deployed first, then the svc. This makes sense as typically the svc will have a dependency on the addons.

For environments, how will the environment addons deployment be sequenced? Before or after the environment is deployed? Or both?

A practical case: If I want to do a AWS WAF, then there's a dependency graph that has to be resolved. It seems there is a race condition.

`AWS::WAFv2::WebACLAssociation` [ref](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-wafv2-webaclassociation.html#aws-resource-wafv2-webaclassociation-syntax) requires a ALB ARN and a WebACL ARN. CloudFormation will solve the dependency between the WebACL and the WebACLAssociation within the stack, but the ALB ARN only exists if the appropriate service exists. Which would only exist after an environment deployment.

I suppose you could make the WebACLAssociation part of the (second) http service addons stack, but the WebACL is more correctly described as being associated to the ALB, not any service or listener rule.

Anything is possible with a little hacking, but I was trying to think of how adding WAF could be done in an easy way --to encourage adoption. Maybe the [CreateALB](https://github.com/aws/copilot-cli/blob/759ccbc3e3fdef8a7a3fe919c20a5d0c696a1e83/internal/pkg/template/templates/environment/cf.yml#L42) CFN condition could be set/forced through a flag in the `<environment>/manifest.yml`. That would assist with the proper association of the WebACL and the ALB.

One final thought that might be beyond the scope of this proposal, but would it make sense for the creation of a higher level abstractions for WAF or AppMesh and link them to properties in the environment manifest? IMO, I think generic addon support is a more valuable, immediate path to pursue, but I just wanted to throw this out there.

Thanks!

**Comment by Lou1415926** [comment]
Thank you @matthewhembree very much for your input! 

> For environments, how will the environment addons deployment be sequenced? Before or after the environment is deployed? Or both?

The short answer: if the env addon depends on some values from the environment (e.g. the WAF use case), the addon will be deployed after. 

---

Below are the details! 

The environment addon stack is deployed as a nested stack, just like a service stack. From CloudFormation's POV, it can be virtually considered as a "resource" (for example, `PublicLoadBalancer` is another resource) to the environment stack. The deployment order of these "resources" (ALB, addon stack, etc.)  are determined by CFN by some dependency graph.

Your service addon is *typically* deployed before the service. This is because your addon typically `Output` something: when Copilot sees an `Output`, it will automatically grab the `Output` and inject it as an envvar/secret/sg/IAMpolicy to your service [[1]](https://aws.github.io/copilot-cli/docs/developing/addons/modeling/#connecting-addon-resources-to-your-workloads). This introduces a dependency on the service addon for your service. CFN recognizes this dependency, and hence deploy the addon before the service.

This is the same with environment addon, with one minor difference - Copilot won't "automatically grab output and inject it [[1]](https://aws.github.io/copilot-cli/docs/developing/addons/modeling/#connecting-addon-resources-to-your-workloads) to your environment". Therefore, as of yet, environment will not have a dependency on the env addon. However, with [customized parameters](https://aws.github.io/copilot-cli/docs/developing/addons/modeling/#customizing-the-parameters-section), the other way around is possible - that your env addon depends on env - in which case, your addon will be deployed after the environment.

For example, if you have [customized parameters](https://aws.github.io/copilot-cli/docs/developing/addons/modeling/#customizing-the-parameters-section) that's like:
```yaml
Parameters:
  ALBARN: !Ref PublicLoadBalancer
```
then the addon will be deployed *after* the environment.


---

> One final thought that might be beyond the scope of this proposal, but would it make sense for the creation of a higher level abstractions for WAF or AppMesh and link them to properties in the environment manifest?

This is a great idea! Thank you ❤️ ! 

I think it makes perfect sense to integrate WAF into environment manifest! 

For AppMesh, there is also a second option: using for example `copilot mesh init` to create an environment addon that defines `AppMesh`. This is similar to how you can run `copilot storage init` to create a service addon template that can be fully configured by yourself.

How would you like this second option in comparison with environment manifest?

**Comment by matthewhembree** [comment]
Very good explanation about stack dependency @Lou1415926 ! That will be very useful to me.

> How would you like this second option in comparison with environment manifest?

I like the idea of using the Copilot CLI as a generator with an `init` verb. That is a good way to provide an opinionated, but extensible implementation.

---
### WAF

On the topic of WAF, I can think of two audiences:
1. The first-time WAF user who needs a good baseline protection without triggering false-positive blocks on their application. An easy way to gain "OWASP Top 10" protection without needing to become too involved in monitoring/logging.
2. An advanced user that might be using something like the [Security Automations for AWS WAF](https://docs.aws.amazon.com/solutions/latest/security-automations-for-aws-waf/welcome.html) or some other custom WebACL.

So maybe one of two possible implementations:
1. In `copilot/environments/<environment>/manifest.yml`:
```yaml
---
# easy mode:
waf:
  enabled: true
---
# custom mode:
waf:
  enabled: true
  webacl: <webacl-arn>
```
2. Using `copilot waf init` to generate an addon template. Possibly a helper flag to associate an an existing WebACL like `--webacl-arn`, but I think it's okay to not have the flag as advanced users could be fine with modifying the generated template. The helper flag might be helpful for larger organizations that have a prebuilt WebACL provided by an infrastructure/security engineer for use by a software engineer to associate to their Copilot environments.

I like the `init` verb generator pattern (option 2) personally, but they're not too different.

---
### AppMesh

As for AppMesh, I think the implementation will be pretty complicated. I don't know if there's work in progress for #1904 , but I, personally, don't think it will be an easy thing to add onto an existing Copilot application. I think the singular goal of mTLS between services sounds easy, but when you need ingress into the mesh, you'll need to implement a virtual gateway. The virtual gateway has a lot of overlap with the existing Copilot ALB architecture --it replaces it. I think AppMesh needs to be decided at the time of `copilot app init`. Otherwise, how would you deal with the mix of meshed and non-meshed services (with TLS `STRICT` mode)? I also think the implementation (of the CFN `Conditions:`) will be simpler at `app init` time.

Another possible complication would be how would Copilot jobs and tasks work in AppMesh? I know in Kubernetes based meshes (Istio (Envoy), Kuma (Envoy), Linkerd2), there are issues with pod termination such that the sidecar container stays running after the job/cronJob main container finishes. Maybe that's a resolved concern in AppMesh with ECS integration.

I don't mean to be so negative on AppMesh. I think it's a valuable feature for many organizations. This is just the first time that I've really thought about the level of effort to implement it as a Copilot orchestration.

**Comment by tmichalski** [comment]
I have nothing to add to the conversation other than I think this is a great addition to the framework. Today our team shoves environment level CFN into the service that "kind of makes sense". Being able to have environment add-ons would more clearly describe the infrastructure visually within our IDE and be more intuitive on where to find the CFN in a larger monorepo.

**Comment by Lou1415926** [comment]
Hello all! The environment addons feature is now released in [v1.25.0](https://github.com/aws/copilot-cli/releases/tag/v1.25.0) 🎉🚀!
